packages/data/src/types.ts

```diff
@@ -1,40 +1,66 @@
-export type WPDataFunctionOrGeneratorArray = Array< Function | Generator >;
-export type WPDataFunctionArray = Array< Function >;
+type MapOf< T > = { [ name: string ]: T };
 
-export interface WPDataAttachedStore {
-	getSelectors: () => WPDataFunctionArray;
-	getActions: () => WPDataFunctionArray;
+export type ActionCreator = Function | Generator;
+export type Resolver = Function | Generator;
+export type Selector = Function;
+
+export type AnyConfig = StoreConfig< any, any >;
+
+export interface StoreInstance< Config extends AnyConfig > {
+	getSelectors: () => SelectorsOf< Config >;
+	getActions: () => ActionCreatorsOf< Config >;
 	subscribe: ( listener: () => void ) => () => void;
 }
 
-export interface WPDataStore {
+export interface StoreDefinition< Config extends AnyConfig > {
```

Comment:
An alternative name that I think would make sense is `StoreDescriptor`. I like the analogy with POSIX file descriptors, where I do `read( fd, buf, bufsize )`, and reading from a store with `registry.select( sd )`. Both descriptors serve as opaque identifiers of the thing to read from.

Reply:
I'm fine with that. I chose "definition" to match the real usage I've seen in the codebase, but as is evident I'm also proposing renames to align those names better to the concepts and roles they fulfill.

I'll change it and if other people have different thoughts they can share them.

Reply:
I've updated this everywhere in 0d8e9b4830